pkg/controllers/pod_controller_test.go

```diff
@@ -0,0 +1,427 @@
+/*
+Copyright 2023 The Kubernetes Authors.
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+    htcp://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package controllers
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sort"
+	"strconv"
+	"testing"
+
+	"github.com/stretchr/testify/assert"
+	batchv1 "k8s.io/api/batch/v1"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/tools/record"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+	"sigs.k8s.io/controller-runtime/pkg/client/interceptor"
+
+	jobset "sigs.k8s.io/jobset/api/jobset/v1alpha2"
+	"sigs.k8s.io/jobset/pkg/constants"
+	testutils "sigs.k8s.io/jobset/pkg/util/testing"
+)
+
+func TestValidatePodPlacements(t *testing.T) {
+	var (
+		jobSetName       = "test-jobset"
+		ns               = "default"
+		leaderPodWrapper = makePod(&makePodArgs{
+			jobSetName:        jobSetName,
+			replicatedJobName: "replicated-job-1",
+			jobName:           "test-jobset-replicated-job-1-test-job-0",
+			podName:           "test-jobset-replicated-job-1-test-job-0-0",
+			ns:                ns,
+			nodeName:          "test-node",
+			jobIdx:            0})
+		podWrapper = makePod(&makePodArgs{
+			jobSetName:        jobSetName,
+			replicatedJobName: "replicated-job-1",
+			jobName:           "test-jobset-replicated-job-1-test-job-0",
+			podName:           "test-jobset-replicated-job-1-test-job-0-1",
+			ns:                ns,
+			jobIdx:            0})
+		nodeSelector = map[string]string{"test-node-topologyKey": "topologyDomain"}
+	)
+	tests := []struct {
+		name           string
+		leaderPod      corev1.Pod
+		podList        *corev1.PodList
+		node           *corev1.Node
+		wantErr        error
+		forceClientErr bool
+		wantMatched    bool
+	}{
+		{
+			name: "topology node label not found",
+			leaderPod: leaderPodWrapper.AddAnnotation(jobset.ExclusiveKey, "test-node-topologyKey").
```

Comment:
Store `"test-node-topologyKey"` in a variable since it is re-used many times in this test. This makes it easier to see what values are supposed to correspond to one another in the tests.